File: 0078-subsets/0078-subsets.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> subsets(int[] nums) {
        int n = nums.length;
        res = new ArrayList<>();
        res.add(new ArrayList<>());
        for(int i=0; i<n; i++) {
            addNumToList(nums[i]);
        }
        return res;
    }
    
    public void addNumToList(int n) {
        int size = res.size();
        for(int i=0; i<size; i++) {
            List<Integer> temp = new ArrayList<>(res.get(i));
            temp.add(n);
            res.add(temp);
        }
    }
}

File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        #backtracking - just form UNIQUE subsets
        result = []
        
        subset = []
        def dfs(i):
            #base case
            if i >= len(nums):
                result.append(subset.copy())
                return
                
            #for all elements, we have 2 options. either to add it or ignore it.
            #choose that element
            subset.append(nums[i])
            dfs(i+1)
            
            #dont choose that element
            subset.pop()
            dfs(i+1)
            
            
        dfs(0)
        return result
                    
                
                